crates/pg_typecheck/src/diagnostics.rs

```diff
@@ -0,0 +1,217 @@
+    hint: Option<String>,
+}
+
+impl Advices for TypecheckAdvices {
+    fn record(&self, visitor: &mut dyn Visit) -> io::Result<()> {
+        // First, show the error code
+        visitor.record_log(
+            LogCategory::Error,
+            &markup! { "Error Code: " <Emphasis>{&self.code}</Emphasis> },
+        )?;
+
+        // Show detailed message if available
+        if let Some(detail) = &self.detail {
+            visitor.record_log(LogCategory::Info, &detail)?;
+        }
+
+        // Show object location information
+        if let (Some(schema), Some(table)) = (&self.schema, &self.table) {
+            let mut location = format!("In table: {schema}.{table}");
+            if let Some(column) = &self.column {
+                location.push_str(&format!(", column: {column}"));
+            }
+            visitor.record_log(LogCategory::Info, &location)?;
+        }
+
+        // Show constraint information
+        if let Some(constraint) = &self.constraint {
+            visitor.record_log(
+                LogCategory::Info,
+                &markup! { "Constraint: " <Emphasis>{constraint}</Emphasis> },
+            )?;
+        }
+
+        // Show data type information
+        if let Some(data_type) = &self.data_type {
+            visitor.record_log(
+                LogCategory::Info,
+                &markup! { "Data type: " <Emphasis>{data_type}</Emphasis> },
+            )?;
+        }
+
+        // Show context information
+        if let Some(where_) = &self.where_ {
+            visitor.record_log(LogCategory::Info, &markup! { "Context:\n"{where_}"" })?;
+        }
+
+        // Show hint if available
+        if let Some(hint) = &self.hint {
+            visitor.record_log(LogCategory::Info, &markup! { "Hint: "{hint}"" })?;
+        }
+
+        // Show source location if available
+        if let (Some(file), Some(line)) = (&self.file, &self.line) {
+            if let Some(routine) = &self.routine {
+                visitor.record_log(
+                    LogCategory::Info,
+                    &markup! { "Source: "{file}":"{line}" in "{routine}"" },
+                )?;
+            } else {
+                visitor.record_log(LogCategory::Info, &markup! { "Source: "{file}":"{line}"" })?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+pub(crate) fn create_type_error(
+    pg_err: &PgDatabaseError,
+    ts: Option<&tree_sitter::Tree>,
+) -> TypecheckDiagnostic {
+    let position = pg_err.position().and_then(|pos| match pos {
+        sqlx::postgres::PgErrorPosition::Original(pos) => Some(pos - 1),
+        _ => None,
+    });
+
+    let range = position.and_then(|pos| {
+        ts.and_then(|tree| {
+            tree.root_node()
+                .named_descendant_for_byte_range(pos, pos)
+                .map(|node| {
+                    TextRange::new(
+                        node.start_byte().try_into().unwrap(),
+                        node.end_byte().try_into().unwrap(),
```

Comment:
I'm not 100% sure, but I think `TextRange::end` is _exclusive_ – maybe we need to add 1 here

Reply:
seems to work fine like this - they are shown at the correct range